golem-worker-executor-base/src/services/worker_fork.rs

```diff
@@ -0,0 +1,178 @@
+use std::marker::PhantomData;
+use std::sync::{Arc, RwLock};
+
+use crate::error::GolemError;
+use crate::metrics::workers::record_worker_call;
+use crate::model::ExecutionStatus;
+use crate::services::oplog::CommitLevel;
+use crate::services::{HasAll, HasOplog};
+use crate::worker::Worker;
+use crate::workerctx::WorkerCtx;
+use async_trait::async_trait;
+use golem_common::model::oplog::OplogIndex;
+use golem_common::model::{
+    AccountId, ComponentType, OwnedWorkerId, Timestamp, WorkerId, WorkerMetadata,
+    WorkerStatusRecord,
+};
+
+#[async_trait]
+pub trait WorkerFork {
+    async fn fork(
+        &self,
+        source_worker_id: &OwnedWorkerId,
+        target_worker_id: &WorkerId,
+        oplog_index_cut_off: OplogIndex,
+    ) -> Result<(), GolemError>;
+}
+
+#[derive(Clone)]
+pub struct DefaultWorkerFork<Ctx: WorkerCtx, Svcs: HasAll<Ctx>> {
+    all: Svcs,
+    ctx: PhantomData<Ctx>,
+}
+
+impl<Ctx: WorkerCtx, Svcs: HasAll<Ctx>> DefaultWorkerFork<Ctx, Svcs> {
+    pub fn new(all: Svcs) -> Self {
+        Self {
+            all,
+            ctx: PhantomData,
+        }
+    }
+
+    async fn validate_worker_forking(
+        &self,
+        account_id: &AccountId,
+        source_worker_id: &WorkerId,
+        target_worker_id: &WorkerId,
+        oplog_index_cut_off: OplogIndex,
+    ) -> Result<(OwnedWorkerId, OwnedWorkerId, WorkerMetadata), GolemError> {
+        let second_index = OplogIndex::INITIAL.next();
+
+        if oplog_index_cut_off < second_index {
+            return Err(GolemError::invalid_request(
+                "oplog_index_cut_off must be at least 2",
+            ));
+        }
+
+        // We assume the target worker is also owned by the same account
+        let owned_target_worker_id = OwnedWorkerId::new(account_id, target_worker_id);
+
+        let target_metadata = self.all.worker_service().get(&owned_target_worker_id).await;
+
+        // We allow forking only if the target worker does not exist
+        if target_metadata.is_some() {
+            return Err(GolemError::worker_already_exists(target_worker_id.clone()));
+        }
+
+        // We assume the source worker belongs to this executor
+        self.all.shard_service().check_worker(source_worker_id)?;
+
+        let owned_source_worker_id = OwnedWorkerId::new(account_id, source_worker_id);
+
+        let metadata = self
+            .all
+            .worker_service()
+            .get(&owned_source_worker_id)
+            .await
+            .ok_or(GolemError::worker_not_found(source_worker_id.clone()))?;
+
+        Ok((owned_source_worker_id, owned_target_worker_id, metadata))
```

Comment:
I think there is a missing `calculate_latest_worker_status` here. The result of `get` is not the latest status just the latest cached one.

(You can say this is a bad API and I agree but let's not change it now :)

Reply:
Sure.

Reply:
I didn't get this part. During forking why would we need the latest worker status of the source worker?

Reply:
Is it to fail the fork if the status is say "exited" or "failed" or "interrupted"?

Reply:
Because you are copying it as the target worker metadata, so I was thinking it is best to copy the latest version and not some random old one :) (Maybe there is no real difference, but easier if you don't have to think it through).

Btw notice later I realized that instead of calculate_latest_worker_status you can get the latest worker status through the worker instance which is better.